src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCM.java

```diff
@@ -348,23 +350,24 @@ public ListBoxModel doFillGitToolItems() {
         @POST
         public ListBoxModel doFillServerIdItems(@QueryParameter String serverId) {
             Jenkins.get().checkPermission(Permission.CONFIGURE);
-            List<BitbucketServerConfiguration> serverList =
-                    new BitbucketPluginConfiguration().getServerList();
+            //Filtered to only include valid server configurations
```

Comment:
Maybe I'm missing something, but if the server with the current ID is invalid, then it still gets returned because of the filter on line 359: `server → server.getId().equals(serverId) || server.validate().kind == FormValidation.Kind.OK` (i.e. if the server id matches, or the server is valid)

Reply:
Yes that's correct. The invalid server is shown if it has been configured. This is my patchy cover for the situation where a server has been reconfigured incorrectly- it's still what the SCM is using but it shouldn't be allowed to be used, so we display it in list but disallow the SCM to be saved until it's changed to a valid one.
The better solution is to remove all references to an invalid server + URL when we save the config, but that belongs to a separate ticket.

Reply:
Right. The comment confused me because it suggested that only valid configurations would be present.